milli/src/update/index_documents/transform.rs

```diff
@@ -340,35 +340,45 @@ impl<'a, 'i> Transform<'a, 'i> {
             return Ok(None);
         }
 
+        // store the keys and values of the json + the original obkv
+        let mut key_value: Vec<(FieldId, Cow<[u8]>)> = Vec::new();
+
+        // the object we're going to use to
         let mut doc = serde_json::Map::new();
 
-        for (k, v) in obkv.iter() {
-            let key = self.fields_ids_map.name(k).ok_or(FieldIdMapMissingEntry::FieldId {
-                field_id: k,
-                process: "Flatten from fields ids map.",
-            })?;
-            let value = serde_json::from_slice::<serde_json::Value>(v)
-                .map_err(crate::error::InternalError::SerdeJson)?;
-            doc.insert(key.to_string(), value);
+        // we recreate a json containing only the fields that needs to be flattened.
+        // all the raw values get inserted directly in the `key_value` vec.
+        for (key, value) in obkv.iter() {
+            if json_depth_checker::should_flatten_from_unchecked_slice(value) {
+                let key = self.fields_ids_map.name(key).ok_or(FieldIdMapMissingEntry::FieldId {
+                    field_id: key,
+                    process: "Flatten from fields ids map.",
+                })?;
+
+                let value = serde_json::from_slice::<serde_json::Value>(value)
```

Comment:
If `serde_json::Value` is already in the scope:

```suggestion
                let value: Value = serde_json::from_slice(value)
```